flash/flash.go

```diff
@@ -0,0 +1,370 @@
+		dev.trans.setClockSpeed(uint32(dev.attrs.MaxClockSpeedMHz) * 1e6)
+	}
+
+	// Enable Quad Mode if available
+	if dev.trans.supportQuadMode() && dev.attrs.QuadEnableBitMask > 0 {
+		// Verify that QSPI mode is enabled.
+		var status byte
+		if dev.attrs.SingleStatusByte {
+			status, err = dev.ReadStatus()
+		} else {
+			status, err = dev.ReadStatus2()
+		}
+		if err != nil {
+			return err
+		}
+		// Check and set the quad enable bit.
+		if status&dev.attrs.QuadEnableBitMask == 0 {
+			if err := dev.WriteEnable(); err != nil {
+				return err
+			}
+			fullStatus := []byte{0x00, dev.attrs.QuadEnableBitMask}
+			if dev.attrs.WriteStatusSplit {
+				err = dev.trans.writeCommand(cmdWriteStatus2, fullStatus[1:])
+			} else if dev.attrs.SingleStatusByte {
+				err = dev.trans.writeCommand(cmdWriteStatus, fullStatus[1:])
+			} else {
+				err = dev.trans.writeCommand(cmdWriteStatus, fullStatus)
+			}
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// disable sector protection if the chip has it
+	if dev.attrs.HasSectorProtection {
+		if err := dev.WriteEnable(); err != nil {
+			return err
+		}
+		if err := dev.trans.writeCommand(cmdWriteStatus, []byte{0x00}); err != nil {
+			return err
+		}
+	}
+
+	// write disable
+	if err := dev.trans.runCommand(cmdWriteDisable); err != nil {
+		return err
+	}
+	return dev.WaitUntilReady()
+}
+
+// Attrs returns the attributes of the device determined from the most recent
+// call to Configure(). If no call to Configure() has been made, this will be
+// the zero value of the Attrs struct.
+func (dev *Device) Attrs() Attrs {
+	return dev.attrs
+}
+
+// ReadJEDEC reads the JEDEC ID from the device; this ID can then be used to
+// ascertain the attributes of the chip from a list of known devices.
+func (dev *Device) ReadJEDEC() (JedecID, error) {
+	jedecID := make([]byte, 3)
+	if err := dev.trans.readCommand(cmdReadJedecID, jedecID); err != nil {
+		return JedecID{}, err
+	}
+	return JedecID{jedecID[0], jedecID[1], jedecID[2]}, nil
+}
+
+// ReadSerialNumber reads the serial numbers from the connected device.
+func (dev *Device) ReadSerialNumber() (SerialNumber, error) {
+	sn := make([]byte, 12)
+	if err := dev.trans.readCommand(0x4B, sn); err != nil {
+		return 0, err
+	}
+	return SerialNumber(uint64(sn[11]) | uint64(sn[10])<<0x8 |
+		uint64(sn[9])<<0x10 | uint64(sn[8])<<0x18 | uint64(sn[7])<<0x20 |
+		uint64(sn[6])<<0x28 | uint64(sn[5])<<0x30 | uint64(sn[4])<<0x38), nil
+}
+
+// ReadBuffer fills the provided buffer with memory read from the device
+// starting at the provided address.
+func (dev *Device) ReadBuffer(addr uint32, buf []byte) error {
+	if err := dev.WaitUntilReady(); err != nil {
+		return err
+	}
+	return dev.trans.readMemory(addr, buf)
+}
+
+// WriteBuffer writes data to the device, one page at a time, starting at the
+// provided address. This method assumes that the destination is already erased.
+func (dev *Device) WriteBuffer(addr uint32, buf []byte) (n int, err error) {
```

Comment:
Same here, it seems to me like `io.WriterAt` would be a good fit here.
The extra requirement of needing to erase first still applies of course, but I don't think that disqualifies SPI flash from implementing the interface.